### DISCUSSION TITLE
What is the secret of automatically using middleware/plugin?

### DISCUSSION THREAD

**Discussion by gjssss** [nuxt/nuxt#24065] [Questions] [answered]
I probably know that it is realized by monitoring the changes of files in the folder, and when files are added, they will be automatically introduced and processed. But I didn't find the relevant code. I found it seems that Vite's fs system was used.

If I want to write a module, it will automatically import the `models` folder in the root directory, and get each exported `export default definition model (...)`  (just like middleware and plugin) in it. How should I realize it?

Can someone give me some key codes, or point out where the related operations are in the nuxt source code?

**Comment by TheAlexLichter** [comment]
Auto importing files !== automatically using middleware or nuxt plugins.

The latter is a tad more complex.
For the first, you can use `imports` in your nuxt config. Check out [my video](https://www.youtube.com/watch?v=FT2LQJ2NvVI), [unimport](https://github.com/unjs/unimport) and [the docs](https://nuxt.com/docs/api/nuxt-config#imports)

**Reply by gjssss** [reply]
Hey! @manniL 
Thank you for your video and explanation. 

However, it seems I might not have explained myself clearly. In fact, what I'm trying to achieve is the latter functionality. 

I'm curious to understand how Nuxt automatically utilizes middleware or plugins because I'd like to implement a similar feature in my modules. 

: >

**Reply by TheAlexLichter** [reply, marked as answer]
Got it!

* Code to initialize and add plugins + middleware to the app [can be found here](https://github.com/nuxt/nuxt/blob/948b30d9c92475d2aa73e5b6f45801361ed9c37d/packages/nuxt/src/core/app.ts#L120-L172)
* [Generating the middleware list](https://github.com/nuxt/nuxt/blob/948b30d9c92475d2aa73e5b6f45801361ed9c37d/packages/nuxt/src/core/templates.ts#L192-L205)
* [Running the middleware](https://github.com/nuxt/nuxt/blob/948b30d9c92475d2aa73e5b6f45801361ed9c37d/packages/nuxt/src/app/plugins/router.ts#L238-L256)
* [Applying plugins](https://github.com/nuxt/nuxt/blob/948b30d9c92475d2aa73e5b6f45801361ed9c37d/packages/nuxt/src/app/entry.ts#L28) + the [function to do so](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/app/nuxt.ts#L333-L347)

Hope that helps!

**Reply by gjssss** [reply]
That is !!! 
Thank you !!!
